Storage/src/Connection/RetryTrait.php

```diff
@@ -0,0 +1,202 @@
+<?php
+/**
+ * Copyright 2022 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Storage\Connection;
+
+/**
+ * Trait which provides helper methods for retry logic.
+ */
+trait RetryTrait
+{
+    /**
+     * @var array
+     */
+    private $httpRetryCodes = [
+        0, // connetion-refused OR connection-reset gives status code of 0
+        408,
+        429,
+        500,
+        502,
+        503,
+        504
+    ];
+
+    /**
+     * @var array
+     */
+    private $opsDescriptionMap = [
+        'bucket_acl.get' => 'idempotent',
+        'bucket_acl.list' => 'idempotent',
+        'buckets.delete' => 'idempotent',
+        'buckets.get' => 'idempotent',
+        'buckets.getIamPolicy' => 'idempotent',
+        'buckets.insert' => 'idempotent',
+        'buckets.list' => 'idempotent',
+        'buckets.lockRetentionPolicy' => 'idempotent',
+        'buckets.testIamPermissions' => 'idempotent',
+        'default_object_acl.get' => 'idempotent',
+        'default_object_acl.list"' => 'idempotent',
+        'hmacKey.delete' => 'idempotent',
+        'hmacKey.get' => 'idempotent',
+        'hmacKey.list' => 'idempotent',
+        'notifications.delete' => 'idempotent',
+        'notifications.get' => 'idempotent',
+        'notifications.list' => 'idempotent',
+        'object_acl.get' => 'idempotent',
+        'object_acl.list' => 'idempotent',
+        'objects.get' => 'idempotent',
+        'objects.list' => 'idempotent',
+        'serviceaccount.get' => 'idempotent',
+        'buckets.patch' => 'idempotent_with_precondition',
+        'buckets.setIamPolicy' => 'idempotent_with_precondition',
+        'buckets.update' => 'idempotent_with_precondition',
+        'hmacKey.update' => 'idempotent_with_precondition',
+        'objects.compose' => 'idempotent_with_precondition',
+        'objects.copy' => 'idempotent_with_precondition',
+        'objects.delete' => 'idempotent_with_precondition',
+        'objects.insert' => 'idempotent_with_precondition',
+        'objects.patch' => 'idempotent_with_precondition',
+        'objects.rewrite' => 'idempotent_with_precondition',
+        'objects.update' => 'idempotent_with_precondition'
+    ];
```

Comment:
Seems like you are missing non-idempotent ops from this, e.g. hmac_key.create.

Reply:
We changed this logic, please have a look now